### PR TITLE
Automatically choose a contrasting axes color

### LINKED ISSUE #15696: Choose legible axes color depending on background color.
The orientation axes, axes grid, and color scale text stays white, even when we change the background color to white. It would be nice to have it choose black or white based on the background color. 

The formula in the link below might be helpful.
http://codepen.io/WebSeed/full/pvgqEq/

### PR DESCRIPTION
**Pull request by quantumsteve**
Description of work.

This adds an option to automatically choose the color of the orientation axes, axes grid, and color scale text so that it will contrast with the background color. This makes it quick to go between a white (for printing) and gray (for viewing on the screen) background.

**To test:**

1. Open the VSI preferences and ensure that "Automatic axes color selection" is checked. If not, enable it.
2. Load a MDWorkspace in the VSI and set the background color to black. Verify that the orientation axes, axes grid, and color scale text are now white.
3. Set the background color to white. Verify that the orientation axes, axes grid, and color scale text are now black.
4. Open the VSI preferences and uncheck "Automatic axes color selection". Restart the VSI.
5. Load a MDWorkspace in the VSI and verify that the axes colors don't vary with the background color.

Fixes #15696.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        helper.Set(value.data(), 3);
+        gridAxis->UpdateProperty(pname);
+      }
+    }
+  }
+}
+
+std::array<double, 3> getContrastingColor(const std::vector<double> &color) {
+  double criteria =
+      1. - (0.299 * color[0] + 0.587 * color[1] + 0.114 * color[2]);
```
Comment (AntonPiccardoSelg) [inline Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp]: What are these numbers?

File: Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+namespace Vates {
+namespace SimpleGui {
+
+namespace {
+
+void safeSetProperty(vtkSMProxy *gridAxis,
+                     std::initializer_list<const char *> pnames,
+                     const std::array<double, 3> &value) {
+  if (gridAxis) {
+    for (auto pname : pnames) {
```
Comment (AntonPiccardoSelg) [inline Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp]: This seems to be copying each element of `pnames`. Is this necessary?

**Comment by AntonPiccardoSelg** [comment]
When I uncheck the `Automatic Axes color selection` though,  it still provides the automatic mechanism. It is only disabled if you restart Mantid. Other than that it works nicely

File: Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp (hunk 2 above)
Comment (quantumsteve) [inline Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp]: It should only copy a pointer of type `const char *`.

File: Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp (hunk 1 above)
Comment (quantumsteve) [inline Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp]: It is a suggested formula for color brightness
https://www.w3.org/TR/AERT#color-contrast

**Comment by quantumsteve** [comment]
@AntonPiccardoSelg Any suggestions on how to change the option without restarting Mantid?

File: Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp (hunk 2 above)
Comment (AntonPiccardoSelg) [inline Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp]: Good point, I didn't see they were pointers

File: Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp (hunk 1 above)
Comment (AntonPiccardoSelg) [inline Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp]: Would it be worth to have that link as a comment?

**Comment by AntonPiccardoSelg** [comment]
@quantumsteve I believe that you should only have to restart the VSI in order to pick up the new settings (and not Mantid). That seems to be the case for other settings such as the background color.

File: Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp (hunk 1 above)
Comment (quantumsteve) [inline Vates/VatesSimpleGui/ViewWidgets/src/VisibleAxesColor.cpp]: I added a comment.

**Comment by quantumsteve** [comment]
@AntonPiccardoSelg The last commit adds a check to remove the observer that was preventing the VSI from changing settings without restarting Mantid.

**Comment by AntonPiccardoSelg** [comment]
@quantumsteve The VSI detects change when enabling `Automatic axes color selection` without having to restart Mantid.  But  when I disable the checkbox, it does not pick that up. Do you know what the reason for this is?

**Comment by quantumsteve** [comment]
@AntonPiccardoSelg It looks like multiple observers were being added but only the last one was deleted. Adding an additional check to see if there is a tag seems to fix the issue you describe above.

**Comment by AntonPiccardoSelg** [comment]
Everything works as described. Please :shipit:

**Pull request merged**